org.eclipse.lsp4e/src/org/eclipse/lsp4e/internal/DocumentOffsetAsyncCache.java

```diff
@@ -0,0 +1,141 @@
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Sebastian Thomschke (Vegard IT GmbH) - initial implementation.
+ *******************************************************************************/
+package org.eclipse.lsp4e.internal;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.Map;
+import java.util.WeakHashMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.IDocumentExtension4;
+
+/**
+ * Generic, per-document+offset cache for asynchronous results that avoids
+ * starting the same work twice by sharing a single running task.
+ *
+ * <p>
+ * Features:
+ * <li>Weakly keys by {@link IDocument} to avoid memory leaks.
+ * <li>Per-document concurrent maps for thread-safe access from UI and
+ * background.
+ * <li>Eviction: TTL-based using {@link System#nanoTime()} and document-change
+ * invalidation when a stable modification stamp is available.
+ * <li>In-flight de-duplication: only one running task per document+offset.
+ * <li>Stale-result protection: if the document changes while a value is being
+ * computed, the result is delivered to callers but is not cached.
+ */
+public final class DocumentOffsetAsyncCache<V> {
+
+	private record Entry<V>(V value, long createdNanos, long docModStamp) {
+		boolean stale(final long ttlNanos, final long currentDocStamp) {
+			return System.nanoTime() - createdNanos > ttlNanos //
+					// Invalidate when we can confidently detect a document change
+					|| (docModStamp != IDocumentExtension4.UNKNOWN_MODIFICATION_STAMP
+							&& currentDocStamp != IDocumentExtension4.UNKNOWN_MODIFICATION_STAMP
+							&& docModStamp != currentDocStamp);
+		}
+	}
+
+	private final Map<IDocument, ConcurrentMap<Integer, Entry<V>>> cache = Collections
+			.synchronizedMap(new WeakHashMap<>());
+	private final Map<IDocument, ConcurrentMap<Integer, CompletableFuture<V>>> inFlight = Collections
+			.synchronizedMap(new WeakHashMap<>());
+
+	private final long ttlNanos;
+
+	public DocumentOffsetAsyncCache(final Duration ttl) {
+		this.ttlNanos = TimeUnit.MILLISECONDS.toNanos(ttl.toMillis());
+	}
+
+	/**
+	 * Returns cached value if present and valid; otherwise returns the single
+	 * running task or starts one via {@code supplier}. A value is valid if it has
+	 * not expired by TTL and (when stamps are available) matches the current
+	 * document stamp. Results computed for an older stamp are not cached.
+	 */
+	public CompletableFuture<V> computeIfAbsent(final IDocument doc, final int offset,
+			final Supplier<CompletableFuture<V>> supplier) {
+		// Fast path: return a completed future if a fresh value is already cached
+		final @Nullable V cachedNow = getNow(doc, offset);
+		if (cachedNow != null)
```

Comment:
in this PR, ifs with only one statement do not have braces around the statement. I think LSP4E has the convention so far to always have braces, doesn't it? Can we have them here as well?

Reply:
AFAIK we do not use braces for single return statement. At least there are plenty of cases like this.